### PR TITLE
Add ConditionalModule

### PR DESCRIPTION
**Pull request by cberner**
This can be used to conditionally bind a module based on a config
property or properties

### CODE REVIEW

File: configuration/src/main/java/io/airlift/configuration/ConditionalModule.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        return new ConditionalModule<>(config, predicate, module);
+    }
+
+    private final Class<T> config;
+    private final Predicate<T> predicate;
+    private final Module module;
+
+    private ConditionalModule(Class<T> config, Predicate<T> predicate, Module module)
+    {
+        this.config = config;
```
Comment (electrum) [inline configuration/src/main/java/io/airlift/configuration/ConditionalModule.java]: Check for null

File: configuration/src/main/java/io/airlift/configuration/ConditionalModule.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * limitations under the License.
+ */
+package io.airlift.configuration;
+
+import com.google.inject.Binder;
+import com.google.inject.Module;
+
+import java.util.function.Predicate;
+
+public class ConditionalModule<T>
```
Comment (electrum) [inline configuration/src/main/java/io/airlift/configuration/ConditionalModule.java]: Add @Beta

File: configuration/src/main/java/io/airlift/configuration/ConditionalModule.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        this.module = module;
+    }
+
+    @Override
+    protected void setup(Binder binder)
+    {
+        T configuration = buildConfigObject(config);
+        if (predicate.test(configuration)) {
+            if (module instanceof ConfigurationAwareModule) {
+                install((ConfigurationAwareModule) module);
```
Comment (electrum) [inline configuration/src/main/java/io/airlift/configuration/ConditionalModule.java]: Can you add a version in the parent that takes a normal Module? The current way of only taking ConditionalModule is confusing and leads to mistakes.

**Comment by cberner** [comment]
Updated

**Comment by electrum** [comment]
Looks good, update CHANGES

**Pull request merged**